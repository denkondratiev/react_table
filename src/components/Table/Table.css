.table {
  cursor: pointer;
  text-align: right;
}
