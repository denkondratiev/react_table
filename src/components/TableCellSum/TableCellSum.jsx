import React from 'react'
import { TableCellSumShape } from '../../helpers/shapes'

const areEqual = (prevProps, nextProps) => {
  const { rowSum } = prevProps

  if (nextProps.rowSum !== rowSum) {
    return false
  }
  return true
}

const TableCellSum = (props) => {
  const {
    id,
    rowSum,
    onMouseEnterPercent,
    onMouseLeavePercent
  } = props

  return (
    <td
      id={id}
      onMouseEnter={onMouseEnterPercent}
      onMouseLeave={onMouseLeavePercent}
    >
      Sum: {rowSum}
    </td>

  )
}

TableCellSum.propTypes = TableCellSumShape.isRequired

export default React.memo(TableCellSum, areEqual)
