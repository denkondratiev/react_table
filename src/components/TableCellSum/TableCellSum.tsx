import React from 'react'
import { TableCellSumProps } from '../../helpers/interface'

const areEqual = (prevProps: TableCellSumProps, nextProps: TableCellSumProps) => {
  const { rowSum } = prevProps

  if (nextProps.rowSum !== rowSum) {
    return false
  }
  return true
}

const TableCellSum: React.FC<TableCellSumProps> = (props) => {
  const {
    rowSum,
    onMouseEnterPercent,
    onMouseLeavePercent
  } = props

  return (
    <td
      onMouseEnter={onMouseEnterPercent}
      onMouseLeave={onMouseLeavePercent}
    >
      Sum: {rowSum}
    </td>

  )
}

export default React.memo(TableCellSum, areEqual)
