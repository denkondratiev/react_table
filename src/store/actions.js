export const ACTION_TYPES = {
  SET_COLUMNS: 'SET::COLUMNS',
  SET_ROWS: 'SET::ROWS',
  SET_TABLE: 'SET::TABLE',
  SET_HIGHLIGHT: 'SET::HIGHLIGHT',
  SET_SHOW_TABLE: 'SET::SHOW::TABLE',
  SET_SHOW_BUTTONS: 'SET::SHOW::BUTTONS'
}

export const setColumns = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_COLUMNS
})

export const setRows = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_ROWS
})

export const setHighlights = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_HIGHLIGHT
})

export const setTable = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_TABLE
})

export const setShowTable = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_SHOW_TABLE
})

export const setShowButtons = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_SHOW_BUTTONS
})
