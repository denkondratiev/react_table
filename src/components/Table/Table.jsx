import React from 'react'
import './Table.css'
import { v4 as uuidv4 } from 'uuid'
import { connect } from 'react-redux'
import { TableShape } from '../../helpers/shapes'
import AverageColumns from '../AverageColumns/AverageColumns'
import { TableRow } from '../TableRow/TableRow'

const Table = (props) => {
  const {
    table,
    showTable
  } = props

  console.log(table)

  return (
    <>
      {showTable && (
        <table className="table table-dark">
          <tbody>
            {table.map((row, rowIndex) => (
              <TableRow
                key={uuidv4()}
                rowIndex={rowIndex}
                row={row}
              />
            ))}
            <AverageColumns />
          </tbody>
        </table>
      )}
    </>
  )
}

Table.propTypes = TableShape.isRequired

const mapStateToProps = (state) => ({
  table: state.table,
  showTable: state.showTable
})

export default connect(mapStateToProps)(Table)
