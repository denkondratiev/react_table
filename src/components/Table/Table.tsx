import React, { useState, SyntheticEvent } from 'react'
import TableRow from '../TableRow/TableRow'
import AverageRow from '../AverageRow/AverageRow'
import { getLightClosest } from '../../helpers/getLightClosest'
import { getRowSum, getLightsAmount, getTable, getRows, getCells } from '../../helpers/selectors'
import { increment } from '../../store/actions'
import './Table.css'
import { useDispatch, useSelector } from 'react-redux'

const Table: React.FC = ()  => {
  const [lightList, setLightList] = useState({})
  const dispatch = useDispatch();

  const table = useSelector(getTable);
  const lightsAmount = useSelector(getLightsAmount);
  const rows = useSelector(getRows);
  const cells = useSelector(getCells);
  const sumRowArray = useSelector(getRowSum)

  const onMouseEnterHandler = (event: SyntheticEvent<HTMLTableCellElement>) => {
    const { id } = event.currentTarget
    const obj = getLightClosest(id, cells, lightsAmount)
    setLightList(obj)
  }

  const onMouseLeaveHandler = (): void => {
    setLightList({})
  }

  const onClickIncrement = (event: SyntheticEvent<HTMLTableCellElement>) => {
    const { id } = event.currentTarget
    dispatch(increment(id))
  }

  return (
    <>
      {table.length > 0 && (
        <table className="table table-dark">
          <tbody>
            {table.map((rowId: string, index: number) => (
              <TableRow
                key={rowId}
                id={rowId}
                rowSum={sumRowArray[index]}
                row={rows[rowId]}
                cells={cells}
                lightList={lightList}
                onMouseEnterHandler={onMouseEnterHandler}
                onMouseLeaveHandler={onMouseLeaveHandler}
                onClickIncrement={onClickIncrement}
              />
            ))}
            <AverageRow />
          </tbody>
        </table>
      )}
    </>
  )
}

export default React.memo(Table)
