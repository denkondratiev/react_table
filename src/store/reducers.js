import { createStore } from 'redux'
import { ACTION_TYPES } from './actions'

function rootReducer (state, action) {
  const {
    SET_COLUMNS,
    SET_ROWS,
    SET_HIGHLIGHTS,
    SET_TABLE,
    SET_SHOW_TABLE,
    SET_SHOW_BUTTONS,
    INCREMENT,
    SET_LIGHT_CELLS
  } = ACTION_TYPES

  switch (action.type) {
    case SET_TABLE: {
      return {
        ...state,
        table: action.payload
      }
    }

    case SET_COLUMNS: {
      return {
        ...state,
        columns: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_ROWS: {
      return {
        ...state,
        rows: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_HIGHLIGHTS: {
      return {
        ...state,
        highlights: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_SHOW_TABLE: {
      return {
        ...state,
        showTable: action.payload
      }
    }

    case SET_SHOW_BUTTONS: {
      return {
        ...state,
        showButtons: action.payload
      }
    }

    case INCREMENT: {
      state.table.splice(action.payload.id, 1, [...state.table[action.payload.id]
        .map((item, index) => index !== action.payload.cellIndex
          ? item
          : {
            ...state.table[action.payload.id][action.payload.cellIndex],
            amount: state.table[action.payload.id][action.payload.cellIndex].amount + 1
          })
      ])
      const newTable = [...state.table]
      return {
        ...state,
        table: newTable
      }
    }

    case SET_LIGHT_CELLS: {
      // const amount = action.payload.amount
      // const highlights = action.payload.highlights
      const rowId = action.payload.id
      const cellIndex = action.payload.cellIndex
      // const col = state.table.flat().length / state.table.length
      console.log(rowId, cellIndex)
      break

      // const lightArray = [...state.table].flat().map((item, index) => ({
      //   ...item,
      //   rowIndex: index,
      //   difference: Math.abs(item.amount - amount)
      // })).sort((a, b) => a.difference - b.difference).slice(0, highlights + 1)

      // state.table.map(cell => (
      //   lightArray.some(lightCell => lightCell.cellId === cell.cellId)
      //     ? { ...cell, isLight: true }
      //     : cell
      // ))

      // return {
      //   ...state,
      //   table: newTable
      // }
    }

    default:
      return state
  }
}

const initialState = {
  rows: '',
  columns: '',
  highlights: '',
  table: [],
  showTable: false,
  showButtons: false
}

const store = createStore(
  rootReducer,
  initialState,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)

export default store
