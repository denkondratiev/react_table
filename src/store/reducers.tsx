import { combineReducers } from 'redux'
import { ACTION_TYPES } from './actions'
import { Action, Params, RowsParams, CellsParams } from './actions'

const paramsReducer = (state: Params = {}, action: Action): Params => {
  const { SET_PARAMS } = ACTION_TYPES

  switch (action.type) {
    case SET_PARAMS:
      return action.payload
    default:
      return state
  }
}

function tableReducer (state: Array<string> = [], action: Action): Array<string> {
  const { SET_TABLE, ADD_ROW, REMOVE_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_TABLE:
      return action.payload
    case ADD_ROW:
      return [...state, action.payload.table]
    case REMOVE_ROW:
      return state.filter(id => id !== action.payload.lastRow)
    default:
      return state
  }
}

const rowsReducer = (state: RowsParams = {}, action: Action): RowsParams => {
  const { SET_ROWS, ADD_ROW, REMOVE_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_ROWS:
      return action.payload
    case ADD_ROW:
      return { ...state, ...action.payload.rows }
    case REMOVE_ROW:
      delete state[action.payload.lastRow]
      return state
    default:
      return state
  }
}


const cellsReducer = (state: CellsParams = {}, action: Action): CellsParams => {
  const { SET_CELLS, INCREMENT, ADD_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_CELLS:
      return action.payload
    case ADD_ROW:
      return { ...state, ...action.payload.cells }
    case INCREMENT:
      return {
        ...state,
        [action.payload]: {
          ...state[action.payload],
          amount: state[action.payload].amount + 1
        }
      }
    default:
      return state
  }
}

const buttonsReducer = (state: boolean = false, action: Action): boolean => {
  const { SHOW_BUTTONS } = ACTION_TYPES

  switch (action.type) {
    case SHOW_BUTTONS:
      return action.payload
    default:
      return state
  }
}

export const rootReducer = combineReducers({
  params: paramsReducer,
  table: tableReducer,
  rows: rowsReducer,
  cells: cellsReducer,
  buttons: buttonsReducer
})
