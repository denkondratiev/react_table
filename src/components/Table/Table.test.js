import React from 'react'
import { create, act } from 'react-test-renderer'
import Table from '../Table/Table'
import TableCellSum from '../TableCellSum/TableCellSum'
import { Provider } from 'react-redux'
import { rootReducer } from '../../store/reducers'
import { createStore } from 'redux'

// START STORE
const params = { rowsAmount: 2, columnsAmount: 2, lightsAmount: 2 }
const table = ['row1', 'row2']
const rows = {
  row1: ['somecell1', 'somecell2'],
  row2: ['somecell3', 'somecell4']
}

const cells = {
  somecell1: { id: 'somecell1', amount: 1 },
  somecell2: { id: 'somecell2', amount: 2 },
  somecell3: { id: 'somecell3', amount: 0 },
  somecell4: { id: 'somecell4', amount: 1 }
}
const buttons = true
const store = createStore(rootReducer, { params, table, rows, cells, buttons })
// END STORE

let component
let root

beforeEach(() => {
  act(() => {
    component = create(
      <Provider store={store}>
        <Table />
      </Provider>
    )
  })
  root = component.root
})

it('test should render initial table', () => {
  expect(component.toJSON()).toMatchSnapshot()
})

it('hover on TableCellSum td with row sum', () => {
  const CellSum = root.findAllByType(TableCellSum)

  act(() => CellSum[0].props.onMouseEnterPercent())

  expect(component.toJSON()).toMatchSnapshot()
})

describe('test cases for td cell', () => {
  const id = 'somecell1'
  const event = { currentTarget: { id } }
  let Cell

  beforeEach(() => {
    Cell = root.findByProps({ id })
  })

  it('hover cell', () => {
    act(() => Cell.props.onMouseEnterHandler(event))

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('leave hover cell', () => {
    act(() => Cell.props.onMouseLeaveHandler())

    expect(component.toJSON).toMatchSnapshot()
  })

  it('click cell', () => {
    act(() => Cell.props.onClickIncrement(event))

    expect(component.toJSON()).toMatchSnapshot()
  })
})
