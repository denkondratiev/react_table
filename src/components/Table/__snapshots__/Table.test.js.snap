// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hover on TableCellSum td with row sum 1`] = `
<table
  className="table table-dark"
>
  <tbody>
    <tr
      id="row1"
    >
      <td
        id="somecell1"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "linear-gradient(90deg, rgba(220,53,69,1) 33.33%, rgba(108,117,125,1) 33.33%)",
          }
        }
      >
        33.33
      </td>
      <td
        id="somecell2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "linear-gradient(90deg, rgba(220,53,69,1) 66.67%, rgba(108,117,125,1) 66.67%)",
          }
        }
      >
        66.67
      </td>
      <td
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Sum: 
        3
      </td>
    </tr>
    <tr
      id="row2"
    >
      <td
        id="somecell3"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "",
          }
        }
      >
        0
      </td>
      <td
        id="somecell4"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "",
          }
        }
      >
        1
      </td>
      <td
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Sum: 
        1
      </td>
    </tr>
    <tr>
      <td>
        Average sum: 
        0
      </td>
      <td>
        Average sum: 
        1
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`test cases for td cell click cell 1`] = `
<table
  className="table table-dark"
>
  <tbody>
    <tr
      id="row1"
    >
      <td
        id="somecell1"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "#6c757d",
          }
        }
      >
        2
      </td>
      <td
        id="somecell2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "",
          }
        }
      >
        2
      </td>
      <td
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Sum: 
        4
      </td>
    </tr>
    <tr
      id="row2"
    >
      <td
        id="somecell3"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "",
          }
        }
      >
        0
      </td>
      <td
        id="somecell4"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "#6c757d",
          }
        }
      >
        1
      </td>
      <td
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Sum: 
        1
      </td>
    </tr>
    <tr>
      <td>
        Average sum: 
        1
      </td>
      <td>
        Average sum: 
        1
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`test cases for td cell hover cell 1`] = `
<table
  className="table table-dark"
>
  <tbody>
    <tr
      id="row1"
    >
      <td
        id="somecell1"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "#6c757d",
          }
        }
      >
        1
      </td>
      <td
        id="somecell2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "",
          }
        }
      >
        2
      </td>
      <td
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Sum: 
        3
      </td>
    </tr>
    <tr
      id="row2"
    >
      <td
        id="somecell3"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "",
          }
        }
      >
        0
      </td>
      <td
        id="somecell4"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "#6c757d",
          }
        }
      >
        1
      </td>
      <td
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Sum: 
        1
      </td>
    </tr>
    <tr>
      <td>
        Average sum: 
        0
      </td>
      <td>
        Average sum: 
        1
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`test cases for td cell leave hover cell 1`] = `[Function]`;

exports[`test should render initial table 1`] = `
<table
  className="table table-dark"
>
  <tbody>
    <tr
      id="row1"
    >
      <td
        id="somecell1"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "",
          }
        }
      >
        1
      </td>
      <td
        id="somecell2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "",
          }
        }
      >
        2
      </td>
      <td
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Sum: 
        3
      </td>
    </tr>
    <tr
      id="row2"
    >
      <td
        id="somecell3"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "",
          }
        }
      >
        0
      </td>
      <td
        id="somecell4"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "background": "",
          }
        }
      >
        1
      </td>
      <td
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Sum: 
        1
      </td>
    </tr>
    <tr>
      <td>
        Average sum: 
        0
      </td>
      <td>
        Average sum: 
        1
      </td>
    </tr>
  </tbody>
</table>
`;
